bolt/lib/Profile/StaleProfileMatching.cpp

```diff
@@ -309,23 +309,28 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
   FlowFunction Func;
 
   // Add a special "dummy" source so that there is always a unique entry point.
-  // Because of the extra source, for all other blocks in FlowFunction it holds
-  // that Block.Index == BB->getIndex() + 1
   FlowBlock EntryBlock;
   EntryBlock.Index = 0;
   Func.Blocks.push_back(EntryBlock);
 
-  // Create FlowBlock for every basic block in the binary function
+  // Create FlowBlock for every basic block in the binary function.
   for (const BinaryBasicBlock *BB : BlockOrder) {
     Func.Blocks.emplace_back();
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    Block.IsExit = BB->successors().empty();
     (void)BB;
     assert(Block.Index == BB->getIndex() + 1 &&
            "incorrectly assigned basic block index");
   }
 
-  // Create FlowJump for each jump between basic blocks in the binary function
+  // Add a special "dummy" sink block so there is always a unique sink.
```

Comment:
I notice there is always special entry block, so there is precedence. But I'm still wondering whether this should be handled inside profi instead? Like alternatively, adding artificial entry/sink can be done inside profi as needed, so this is all transparent to the users of profi.

Reply:
In order to construct the artificial sink block and edges from exit blocks to the sink block, BinaryBasicBlock information is/will be needed, which is not currently exposed to SampleProfileInference/profi. For instance, identifying calls to functions that pass control back outside of the function (i.e. calls to std::terminate) requires BinaryBasicBlock information. (Identifying these calls is something I plan to implement in a subsequent commit). I think it also makes sense organizationally to do this in the construction of the flow function - it would be inelegant to expose BinaryBasicBlock information to profi (code duplicity) just to add the sink block. The construction of the entry block could be more easily moved to profi, but given that sink block should be constructed within createFlowFunction it makes sense for that block's construction to remain where it is as well.

Reply:
I don't see why do we need BinaryBasicBlock info in profi. So we construct a CFG represented with flow blocks first. BinaryBasicBlock is only used during CFG construction like it is now. Then within profi, if we find multiple exit blocks (no successors) in that CFG, we just connect these exit blocks into a dummy sink block, without needing info from original BinaryBasicBlock.

Reply:
We would need binary block info because the condition "block has no successors" does not fully encompass our definition of "exit blocks". For instance, there are landing pads flow blocks that have cfg successors but call/transitively std::terminate or some other function that transfers control back outside of the function, thus qualifying it as an exit block. We need binary block information to parse through instructions and identify these calls. I plan to cover these degenerate cases in a following commit.

Reply:
> "block has no successors" does not fully encompass our definition of "exit blocks"

What is our definition of "exit blocks"? A reasonable definition could be a block that transfer control outside of the function. Now successor blocks naturally implies the successor within the same function. With that definition, blocks containing no return calls should not have successors, and such information should be embeded in the CFG already (`isExitBlock`), without needing to look at BinaryBasicBlock again. If we do need extra info to know where we have exit flow due to noreturn call after CFG is constructed, that indicates the CFG construction is wrong? 

> I plan to cover these degenerate cases in a following commit. 

Evidently, the way this changes are split causes confusion. This patch as it is doesn't make its intention obvious. It would be better to just include those intended changes in one patch.

Reply:
> With that definition, blocks containing no return calls should not have successors,
I don’t think this is universally true for exceptions. Consider a block that has a return instruction and throws an exception. The BBB representation would report that there are no successors, but in our current construction of the flow CFG, we would not label that as an exit block – the landing pad would be considered a successor and we’d lose out on applying profile inference to that function. However, this new implementation “ignores” landing pads in the identification of exit blocks which increase our scope (the added test reflects this). I took your advice on IsExit and ExitBlockIndicies – those were redundant as you pointed out and sorry about the confusion.

Reply:
Had an offline discussion with @shawbyoung, now I see the problem (thanks for explaining this with actual CFG). So a real exit block can have exception, which makes landing pad its successor, and since landing can recover from exception and continue execution, it will go back to the function. Based on our discussions, there are a few things we need to do:

1) this issue shouldn't be unique to BOLT, worth checking how compiler is handling it. anytime we do something specific for compiler or bolt, if feels fishy because their need should be similar and it should be handled in infer without duplication.

2) we actually need to tell profi that exception flow is rare and it should take that into consideration when doing inference. one simple approximation is just to remove edges to landing pad altogether so profi won't send flow to exception path which is supposed to be very very rare anyways. doing so also solves the problem without adding sink node. However it looks like from the compiler side, we communicate such constraints via "unlikely" flag on edge, so profi can adjust edge costs. So we could also do the same for bolt. One way or another we need to profi to take it into consideration. 

It's okay if we leverage unlikely flag for EH path in a follow up patch, but we probably want an answer for #1 here.

Reply:
The case handled by this PR (exit block, e.g. a block that ends with return/tail call, containing an exception throwing function call) is unique to BOLT. The compiler maintains the invariant "Invoke ends the block” while BOLT treats Invokes as regular calls that don't terminate the block, and merges the fallthrough block which might be an exit block. Work should be done to mirror this in BOLT - splitting blocks with exception throwing calls at the call and adding the landing pad as the successor. However, this change would be extensive, and this PR is a relatively speaking minimally invasive change that is functionally equivalent.

Reply:
Sounds good. Agreed that splitting the flow from invoke is more natural. 

Can you add a comment document this as a TODO, so we know why there is divergence in how we handle flow to landing pad? Otherwise this patch looks good. Also helpful to have a TODO comment for unlikely jump flag to communicate to profi. 

Thanks for working through the feedbacks.